### ISSUE TITLE
D7 -> D8 File migration

### ISSUE THREAD

**Issue by cory-skeers** [uiowa/uiowa#379]
### Proposed Feature
Create a file migration option for moving sites from D7 to D8.

Content migration is likely out of scope, due to changes with paragraphs, different content types for different sites, etc., but an automatic file migration could help save time.

D8 migration is part of core, including D2D migrations:
https://www.drupal.org/docs/8/api/migrate-api/migrate-api-overview

Specific field mappings need to be customized, but can likely be done in a manner that would work for most if not all SiteNow websites.

**Comment by joewhitsitt** [comment]
To get the most out of this migration I think this should be expanded to create managed file and the associated media entity.

**Comment by joewhitsitt** [comment]
I saw some Drupal Slack chatter about this, though I haven't looked at this:
https://www.drupal.org/sandbox/srjosh/3062884

**Comment by cory-skeers** [comment]
Interesting... From a quick look through, it looks like their MediaMaker.php plugin would handle most/all of the destination portion of the migration. 

Looking at the source half of it, looks like
- It loads every page, pulls in-line images for their sources, then downloads
- Uploads these images using the destination portion of the migration

I think it'd be more efficient to sync the files and then query the database directly, where you could pull all file information with a single query, but on an individual site level, might be more efficient to not reinvent the wheel if this works.

For the inline media, I have the base code for a WordPress -> D7 migration written from another project, but for it and this implementation, it assumes a single body content field. It could work for a straight migration, but either each page could only have one section/text bundle in its Paragraphs.

### To do
- Check destination mapping in https://www.drupal.org/sandbox/srjosh/3062884 and/or update its MediaMaker.php
- Check for a sync/query option in place of its individual node load approach
- Decide if any nodes and/or text will be migrated (and then their inline-embeds, and image fields), or files only

**Comment by cory-skeers** [comment]
New updates to https://www.drupal.org/project/migrate_media_handler. Need to review/test, but it may be close to or fully an option now.

**Comment by cory-skeers** [comment]
### Helpful modules
https://www.drupal.org/project/migrate_upgrade
https://www.drupal.org/project/migrate_tools

These work better out of the box than expected. Migrating from D7 to D8 databases goes smoothly. 

### File migration
files_managed migrates correctly. Additional 'type' designation needs to be added to files to map to the correct media entity bundle, but this can be based off the MediaMaker.php (https://www.drupal.org/sandbox/srjosh/3062884)

### Users / Taxonomies
Users and taxonomies can be migrated as well, possibly aiding in migrating custom sites, though user roles may get thorny.

### Nodes / Content
For custom D7 sites with Paragraphs, custom content types (like locations/events) that need to be recreated, or content types which do not use Paragraphs, may be able to directly migrate content with no custom work.

For other cases, such as D7 pages without Paragraphs migrating to D8 pages with Paragraphs, custom work needed, but may be straightforward--put all content into a single Paragraphs section/textarea to get the pages built and content in the site, and fine-tuning and adjustments can be handled manually after.

**Comment by cory-skeers** [comment]
File/Image (and article) migrations are functional for SiteNow Standard sites.

### To do

- Basic Page migration (in progress)
  - Paragraphs migration first, then associate with node migration
  - Paragraphs (and article content) need image replacement in prepareRow() to update to new media entities
- Database config
  - Sync from local, or option of remote w/aliases
- Improve rollback feature
  - Secondary elements (like media entities during file migration) are not removed from D8 database on migration rollback, and can lead to errors)
- Error catching
  - Migrations not robust to anomalies, unexpected file types, etc. Combined with lack of rollback ties can lead to issues

### "Would be nice"

- File copy direct from database, either remote or after rsync to local
- Migration manifest yml
- Script to automate process
- Taxonomy migrations

**Comment by cory-skeers** [comment]
File/Image, Article, and Basic Page implementation in place.

## To test

- Locally provision a D7 site.
- Run Dev Desktop
- Update docroot/modules/custom/sitenow_migrate/config/install/migrate_plus.migration_group.sitenow_migrate.yml
source:
  database:
    database: **mysite_databasename**
- Update docroot/modules/custom/sitenow_migrate/config/install/migrate_plus.migration.d7_file.yml
source:
  plugin: files
  constants:
    SOURCE_BASE_PATH: **"http://baseurlpathforfiles"** (dev/stage can be used if the files are up-to-date)
-  Run the following (d7_file must be imported before d7_article or d7_page):
`drush en sitenow_migrate`
`drush mim d7_file`
`drush mim d7_article`
`drush mim d7_page`

Note: drupal/migrate_plus and drupal/migrate_tools are needed.
Migrate config files are created on install, and if any changes are needed, can 
`drush pmu sitenow_migrate` 
and rerun the above script--the module will delete the existing config files on uninstall so they can be recreated on subsequent installs.

Other useful migrate commands:
`drush ms` : Show registered migrations and number of imports
`drush migrate-stop && drush migrate-reset` : Stop and reset the status for a migration that hit a breaking error
`drush mr --all` : Rollback (remove) all imported content (this will remove all content **except** for media entities, which will remain in the database unless manually removed).

**Comment by cory-skeers** [comment]
@uiowa/its-web -- Migration module is available in the migrate branch. Ready for any testing/additions/feedback.

Instructions are available in https://github.com/uiowa/uiowa/issues/379#issuecomment-533297141.

**Should** work for standard sites in general, and to some extent custom sites (won't work for custom content types or panels, for instance, but files and basic pages can be imported).

Further migrations can be set for taxonomies if the need arises, and we can look into migrations for content types such as Panels, but more discussion would be needed first.

### Known Issues
~~HTML tags will display in the body on import. This is from the migration not handling the filtered_html body/format due to a buggy sub_process plugin. Current workaround is to open the imported page and save (no edits are necessary--opening and saving will update the format from NULL to filtered_html).~~ _This has been fixed in the PostRow event listener for both basic pages and articles._

Files with duplicate names are migrated with a _# appended to their filename. Because filename is used to track files across migration, this results in any inline file embeds to use only the first file. (E.g. if there are three different logo.png files, they will import as logo.png, logo_1.png, and logo_2.png, but all instances of body embeds will use only logo.png).

All images are pulled in as `image` and are not directly useable by `image_background` fields. This could be done post-migration by manually downloading the file and re uploading as an image_background, or (potentially) by updating the database directly--moving from `media__field_media_image` to `media__field_media_image_1`

**Comment by joewhitsitt** [comment]
Just took a look at this. AWESOME! I didn't use a very good candidate site, but the file migration alone was so speedy. I am going to try another site here soon because this is pretty exciting.

I got the branch up to date with master and started a draft PR #594 

Initial thoughts:
We might be able to move the config into the sitenow sync dir. Set some defaults for those two config changes. Ignore those files in sitenow. Turn on these modules by default. Create a admin UI to modify the database credentials and files path. I think it would be beneficial to still run the commands from the command line so that we can easily see the errors to preform any follow-up. All this combined might keep us from having to make a site "custom" to run this stuff.

Taxonomy tags for articles would be a nice addition.

Sounds like Cory is working on the person content type as well.

This is another instance of why we should probably do away with the image_background media type through an update hook. It has created unnecessary complication for little benefit. I created #596 to possibly address this.

Ultimately, as-is, I see this as a great offering for Standard sites that weren't going to get anything, and a great base for working on custom sites.
